python/cudf/cudf/io/json.py

```diff
@@ -162,12 +162,47 @@ def read_json(
 
 
 @ioutils.doc_to_json()
-def to_json(cudf_val, path_or_buf=None, *args, **kwargs):
+def to_json(
+    cudf_val,
+    path_or_buf=None,
+    engine="auto",
+    storage_options=None,
+    *args,
+    **kwargs,
+):
     """{docstring}"""
 
-    warnings.warn(
-        "Using CPU via Pandas to write JSON dataset, this may "
-        "be GPU accelerated in the future"
-    )
+    if engine == "cudf":
+        return_as_string = False
+        if path_or_buf is None:
+            path_or_buf = StringIO()
+            return_as_string = True
+        path_or_buf = ioutils.get_writer_filepath_or_buffer(
+            path_or_data=path_or_buf, mode="w", storage_options=storage_options
+        )
+        orient = kwargs.pop("orient", "records")
+        if orient != "records":
+            raise ValueError(
+                "Only the 'records' orient is supported for cudf JSON writer"
+            )
+
+        if ioutils.is_fsspec_open_file(path_or_buf):
+            with path_or_buf as file_obj:
+                file_obj = ioutils.get_IOBase_writer(file_obj)
+                libjson.write_json(
+                    cudf_val, path_or_buf=file_obj, *args, **kwargs
+                )
+        else:
+            libjson.write_json(
+                cudf_val, path_or_buf=path_or_buf, *args, **kwargs
+            )
+
+        if return_as_string:
+            path_or_buf.seek(0)
+            return path_or_buf.read()
+        return
+    warnings.warn("Using CPU via Pandas to write JSON dataset")
     pd_value = cudf_val.to_pandas(nullable=True)
-    return pd.io.json.to_json(path_or_buf, pd_value, *args, **kwargs)
+    return pd.io.json.to_json(
+        path_or_buf, pd_value, *args, storage_options=storage_options, **kwargs
+    )
```

Comment:
This can be re-written as:
```python
    if engine == "auto":
        engine = "pandas" if orient is not None and orient != "records" else "cudf"

    if engine == "cudf":
        if orient not in {"records", None}:
            raise ValueError(
                f"Only the `orient='records'` is supported for JSON writer with `engine='cudf'`, got {orient}"
            )

        if path_or_buf is None:
            path_or_buf = StringIO()
            return_as_string = True
        else:
            path_or_buf = ioutils.get_writer_filepath_or_buffer(
                path_or_data=path_or_buf, mode="w", storage_options=storage_options
            )
            return_as_string = False

        if ioutils.is_fsspec_open_file(path_or_buf):
            with path_or_buf as file_obj:
                file_obj = ioutils.get_IOBase_writer(file_obj)
                libjson.write_json(
                    cudf_val, path_or_buf=file_obj, *args, **kwargs
                )
        else:
            libjson.write_json(
                cudf_val, path_or_buf=path_or_buf, *args, **kwargs
            )

        if return_as_string:
            path_or_buf.seek(0)
            return path_or_buf.read()
    else:
        warnings.warn("Using CPU via Pandas to write JSON dataset")
        pd_value = cudf_val.to_pandas(nullable=True)
        return pd.io.json.to_json(
            path_or_buf, pd_value, storage_options=storage_options, *args, **kwargs
        )
```

Reply:
Also, can we default `engine=auto` to `engine=pandas`? Like, unconditionally defaulting to pandas engine when `auto` is set:
```python
     if engine == "auto":
        engine = "pandas"
```

The reason I ask this we seem to have som breakages in pytests when I tested out default to using `cudf` engine and pandas seems to be unable to read/parse our json output:
```bash
FAILED python/cudf/cudf/tests/test_json.py::test_json_writer[0] - AssertionError: DataFrame are different
FAILED python/cudf/cudf/tests/test_json.py::test_json_writer[1] - AssertionError: Attributes of DataFrame.iloc[:, 10] (column name="col_datetime64[ms]") are different
FAILED python/cudf/cudf/tests/test_json.py::test_json_writer[10] - ValueError: Unexpected character found when decoding array value (2)
FAILED python/cudf/cudf/tests/test_json.py::test_json_writer[100] - ValueError: Unexpected character found when decoding array value (2)
```

There are a bunch of other errors but those all can be fixed with test refactoring. But the above seems to be the critical ones.

Reply:
cc: @vuule What do you think?

Reply:
>  can we default engine=auto to engine=pandas?

Sounds good, the writer needs to be opt-in for now.

Reply:
> related: The engine parameter needs to be added to docs, and IMO "cudf" engine should be documented as experimental for now.

+1

Reply:
added `engine` to docs, defaulted `auto` to `pandas`.